### PR TITLE
[TRTLLM-13319][fix] fix output distribution correctness for Eagle3 dynamic-tree rejection sampling

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Summary by CodeRabbit

# Release Notes

* **Performance Improvements**
  * Optimized speculative decoding verification path with simplified computation logic, reducing unnecessary intermediate steps and memory overhead.
  * Added compilation optimization for probability calculations to improve inference throughput.

* **Refactor**
  * Streamlined internal verification interface for speculative decoding, removing intermediate computation stages while maintaining accuracy.

## Summary

This PR fixes the probability distribution used during verification in Eagle3 one-model dynamic tree rejection sampling, and includes a refactor and a new feature.

## Changes

- **Fix**: Correct probability distribution in dynamic tree rejection sampling — previously using incorrect logits slice for verification
- **Refactor**: Pass `draftTokens[num_gens, N-1]` directly to rejection kernel, removing redundant reshaping
- **Feat**: Support `TLLM_SPEC_DECODE_FORCE_NUM_ACCEPTED_TOKENS` for both dynamic tree and linear rejection sampling paths. Previously unsupported because rejection sampling processes tokens sequentially; when the forced length exceeds actual accepted length, unfilled slots contain out-of-vocabulary garbage. Fix: pre-fill those positions with draft tokens when force mode is active.

## Test Coverage

> TRT-LLM numbers reflect batch-full periods only (all concurrency slots occupied), so they appear higher than system-level throughput.

| Model | Framework | greedy tok/s | rejection tok/s | AL (greedy) | AL (rejection) | AR (greedy) | AR (rejection) |
|---|---|---|---|---|---|---|---|
| Qwen3-8B | SGLang | 2060| 1998 | 3.23 | 3.18 | - | - |
| Qwen3-8B | TRT-LLM | 2977.2 | 2806.6 | 3.327 | 3.330 | 0.776 | 0.777 |
| Qwen3-235B-FP8 | SGLang | 977| 1022 | 2.33 | 2.34 | - | - |
| Qwen3-235B-A22B | TRT-LLM | 1344.7 | 1332.3 | 2.713 | 2.709 | 0.571 | 0.570 |
| Llama-3.3-70B | SGLang | 1381| 1354 | 2.33 | 2.31 | - | - |
| Llama-3.3-70B | TRT-LLM | 1168.3 | 1118.6 | 3.747 | 3.708 | 0.458 | 0.451 |

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by zhaoyangwang-nvidia** [comment]
Hi @NVIDIA/trt-llm-torch-graph-compiler @NVIDIA/trt-llm-torch-spec-decoding please help to review this PR, thanks a lot~

File: tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py [hunk 1]
```diff
@@ -834,115 +820,145 @@ def _sample_and_accept_dynamic_tree(
...
                 )
+                num_accepted_tokens[num_contexts:batch_size] = 1
+                accepted_tokens[num_contexts:batch_size, 0] = gen_root_tokens
+                self._accepted_draft_indices_tensor[num_contexts:batch_size] = -1
                 return accepted_tokens, num_accepted_tokens
 
+            target_logits_tree = logits[-num_gens * N :].reshape(-1, vocab_size)
+            gen_slice = slice(num_contexts, num_contexts + num_gens)
+            temps = spec_metadata.request_temperatures[gen_slice]
+            top_ks_rej = spec_metadata.request_top_ks[gen_slice]
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py:914]: Do you have to clamp these like on the context path? I think things can go wrong if they're set to 0 (e.g. topk disabled)

File: tensorrt_llm/_torch/speculative/one_model_sampler.py [hunk 2]
```diff
@@ -117,6 +117,7 @@ def sampling_batch_spec_dec_one_model(
     return random_sampled
 
 
+@torch.compile(options={"max-autotune": True})
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/speculative/one_model_sampler.py:135]: Can you confirm that this doesn't cause issue during warmup for the existing linear chain path? torch.compile during stream capture can fail

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/one_model_sampler.py:135]: Confirmed empirically on H200 with Qwen3-8B Eagle3 (one-model linear chain, rejection sampling on, CUDA graph on, non-greedy temp/top_k/top_p). Warmup + capture complete cleanly (exit 0, full throughput results); the dedicated advanced sampling capture pass exercises compute_probs_from_logits for all 4 captured batch sizes with no stream is capturing / inductor / dynamo errors. The capture path runs WARMUP_STEPS eager iterations of the exact callable before torch.cuda.graph(), so max-autotune compilation/autotuning happens outside the capture region. No capture failure.

File: tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py (hunk 1 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py:914]: Good catch — this turned out to be worth digging into. Here's what we changed and why:

1. Moved the clamp into the sampling helpers instead of the call site. The top-k range requirement really belongs to the functions that call flashinfer (compute_probs_from_logits / sampling_batch_spec_dec_one_model), not to each caller. Keeping it at the call site is exactly how this bug slipped in — the context path clamped but the rejection path didn't. Centralizing it in a shared sanitize_top_k() means every current and future caller is protected automatically and we can't forget it again.

2. Upper bound is now vocab_size, not vocab_size - 1. Checking the flashinfer kernel (topk.cuh), top_k_mask_logits only takes the "no masking, just copy" fast path when k >= vocab_size. With vocab_size - 1 it still runs a full radix top-k selection (just to drop the single lowest-logit token), which is wasted work for a request that wants no top-k filtering. Clamping to vocab_size lets the disabled case hit the skip path. I verified k = vocab_size is safe on all three backends (flashinfer mask + fused sampling op, the C++ op, and the PyTorch-native path all keep-all without error).

3. Disabled top-k must map to vocab_size, not 1. Per SamplingParams, top_k = 0 (and the effective default when sampling is on) means "all logits" — i.e. top-k disabled. The problem is our three operators don't agree on a literal 0: the C++ op treats 0 as keep-all, but flashinfer masks the entire row (all-zero probs) and the PyTorch-native path gathers out of bounds. So clamping 0 → 1 would be wrong (that's greedy, the opposite of "all logits"). The new normalization mirrors the C++ op — where(k > 0, k, vocab_size) — so 0 correctly becomes keep-all across all backends.

4. Scope. This fixes the top-k inconsistency for now. I suspect other sampling params (top-p, temperature) may have similar "param definition vs. operator behavior" mismatches, so I'll audit the full set and fix them properly as part of the upcoming sampler refactor rather than patching each one here.

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py [hunk 3]
```diff
@@ -1247,63 +1247,10 @@ def _(logits: torch.Tensor,
...
-    @torch.library.register_fake("trtllm::build_draft_prob_indices_out_op")
-    def _(topkScoreIndices: torch.Tensor, draftProbIndices: torch.Tensor,
-          topK: int, numDraftTokens: int) -> None:
-        return None
-
     @torch.library.register_fake("trtllm::verify_dynamic_tree_rejection_out_op")
-    def _(candidates: torch.Tensor, draftProbs: torch.Tensor,
-          targetProbs: torch.Tensor, targetSupportIndices: torch.Tensor,
-          targetSupportLengths: torch.Tensor, draftProbIndices: torch.Tensor,
+    def _(draftTokens: torch.Tensor, targetProbs: torch.Tensor,
```
Comment (liji-nv) [inline tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py]: This fake impl is unnecessary. As it is an inplace op. Better adding to inplace_info() in tensorrt_llm/_torch/compilation/utils.py

**Review by liji-nv — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py (hunk 3 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py]: Fixed

**Comment by zhaoyangwang-nvidia** [comment]
Hi @mikeiovine , I’ve addressed all the comments you left earlier and made the corresponding changes. Could you please take another look when you have a chance? Thanks!

**Review by mikeiovine — APPROVED** [review]
